src/keyrefresh/participant.rs

```diff
@@ -0,0 +1,1065 @@
+    }
+
+    /// Handle the protocol's round three public messages.
+    ///
+    /// Here we validate the Schnorr proofs from each participant.
+    #[cfg_attr(feature = "flame_it", flame("keyrefresh"))]
+    #[instrument(skip_all, err(Debug))]
+    fn handle_round_three_msg(
+        &mut self,
+        message: &Message,
+    ) -> Result<ProcessOutcome<<Self as ProtocolParticipant>::Output>> {
+        info!("Handling round three keyrefresh message.");
+
+        if !self.local_storage.contains::<storage::GlobalRid>(self.id()) {
+            self.stash_message(message)?;
+            return Ok(ProcessOutcome::Incomplete);
+        }
+        let global_rid = *self
+            .local_storage
+            .retrieve::<storage::GlobalRid>(self.id())?;
+
+        let proofs = PiSchProof::from_message_multi(message)?;
+        let decom = self
+            .local_storage
+            .retrieve::<storage::Decommit>(message.from())?;
+
+        // Check that there is one proof per participant.
+        if proofs.len() != self.all_participants().len() {
+            error!("Received incorrect number of proofs",);
+            return Err(InternalError::ProtocolError(Some(message.from())));
+        }
+
+        for ((proof, precommit), update_public) in proofs
+            .into_iter()
+            .zip(decom.As.iter())
+            .zip(decom.update_publics.iter())
+        {
+            let mut transcript = schnorr_proof_transcript(self.sid(), &global_rid, message.from())?;
+            proof.verify_with_precommit(
+                CommonInput::new(update_public),
+                &self.retrieve_context(),
+                &mut transcript,
+                precommit,
+            )?;
+        }
+
+        // Only if the proof verifies do we store the participant's updates.
+        self.local_storage
+            .store::<storage::ValidPublicUpdates>(message.from(), decom.update_publics.clone());
+
+        self.maybe_finish()
+    }
+
+    /// Handle the protocol's round three private messages.
+    ///
+    /// Here we validate and store a private update from someone to us.
+    #[cfg_attr(feature = "flame_it", flame("keyrefresh"))]
+    #[instrument(skip_all, err(Debug))]
+    fn handle_round_three_msg_private(
+        &mut self,
+        message: &Message,
+    ) -> Result<ProcessOutcome<<Self as ProtocolParticipant>::Output>> {
+        info!("Handling round three keyrefresh private message.");
+
+        message.check_type(MessageType::Keyrefresh(
+            KeyrefreshMessageType::R3PrivateUpdate,
+        ))?;
+        let encrypted_update: KeyUpdateEncrypted = deserialize!(&message.unverified_bytes)?;
+
+        // Get my private key from the AuxInfo protocol.
+        let my_dk = self.input.private_auxinfo().decryption_key();
+
+        // Decrypt the private update.
+        let update_private = encrypted_update.decrypt(my_dk)?;
+
+        // TODO: check that this private share matches our public share in decom.update_publics.
```

Comment:
Should something be done about this TODO?

Reply:
Done.